python/lsst/daf/butler/registry/interfaces/_versioning.py

```diff
@@ -0,0 +1,195 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = [
+    "VersionTuple",
+    "VersionedExtension",
+]
+
+from abc import ABC, abstractmethod
+import hashlib
+from typing import (
+    Iterable,
+    NamedTuple,
+    Optional,
+)
+
+import sqlalchemy
+
+
+class VersionTuple(NamedTuple):
+    """Class representing a version number.
+
+    Parameters
+    ----------
+    major, minor, patch : `int`
+        Version number components
+    """
+    major: int
+    minor: int
+    patch: int
+
+    @classmethod
+    def fromString(cls, versionStr: str) -> VersionTuple:
+        """Extract version number from a string.
+
+        Parameters
+        ----------
+        versionStr : `str`
+            Version number in string form "X.Y.Z", all components must be
+            present.
+
+        Returns
+        -------
+        version : `VersionTuple`
+            Parsed version tuple.
+
+        Raises
+        ------
+        ValueError
+            Raised if string has an invalid format.
+        """
+        try:
+            version = tuple(int(v) for v in versionStr.split("."))
+        except ValueError as exc:
+            raise ValueError(f"Invalid version  string '{versionStr}'") from exc
+        if len(version) != 3:
+            raise ValueError(f"Invalid version  string '{versionStr}', must consist of three numbers")
+        return cls(*version)
+
+    def __str__(self) -> str:
+        """Transform version tuple into a canonical string form.
+        """
+        return f"{self.major}.{self.minor}.{self.patch}"
+
+
+class VersionedExtension(ABC):
+    """Interface for extension classes with versions.
+    """
+
+    @classmethod
+    @abstractmethod
+    def currentVersion(cls) -> Optional[VersionTuple]:
+        """Return extension version as defined by current implementation.
+
+        This method can return ``None`` if an extension does not require
+        its version to be saved or checked.
+
+        Returns
+        -------
+        version : `VersionTuple`
+            Current extension version or ``None``.
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def extensionName(cls) -> str:
+        """Return full name of the extension.
+
+        This name should match the name defined in registry configuration. It
+        is also stored in registry attributes. Default implementation returns
+        full class name.
+
+        Returns
+        -------
+        name : `str`
+            Full extension name.
+        """
+        return f"{cls.__module__}.{cls.__name__}"
+
+    @abstractmethod
+    def schemaDigest(self) -> Optional[str]:
+        """Return digest for schema piece managed by this extension.
+
+        Returns
+        -------
+        digest : `str` or `None`
```

Comment:
My concern here was that we need to save them in the database as strings because attribute table is a string -> string mapping. We could require bytes digests but then I'd need to convert them into string format (e.g. with hex encoding) when saving and convert back to bytes when reading. I decided to leave the choice for encoding to extension, besides there may be a case when digest can be produced as an actual human-readable string.